### PR TITLE
Replace Jest with Vitest 2

### PR DESCRIPTION
**Pull request by birkskyum**
This PR replaces the test runner (unit/integration/build) from Jest to [Vitest 2](https://vitest.dev/), which has built-in support for TS/ESM and the same syntax as Jest. This PR serves to help us evaluate the option, and to move forward faster if there's a decision to proceed at some point.

A key motivation for examining this change is that I've found it quite challenging to move these tickets forward:
- #1595 
- #977 

Part of the struggle is that Jest only has [experimental ESM support](https://jestjs.io/docs/ecmascript-modules), and I've so far not managed to see an effect of the [useESM](https://kulshekhar.github.io/ts-jest/docs/next/guides/esm-support/) setting in the ts-jest plugin used in this repo.

It's a continuation of 
- #4684 

Relates to Rolldown, because vitest is downstream of vite, and thus eventually will swap internals to use rolldown:
- #4669 

Relates a bit to browser runner, because vitest is working on a browser mode:
- #4949 

## Launch Checklist
 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!

### CODE REVIEW

**Comment by jasongitmail** [comment]
@birkskyum  I took a look at the unit test config as you requested. 

Looks good afaict. Coverage is in a separate command, `fileParallelism` is left at the default (true), etc.

If there is a particular file that consistently runs more slowly, using [`describe.concurrent()` within it could help](https://vitest.dev/api/#describe-concurrent). And I'd carefully look at how `beforeEach()` is used and consider if `beforeAll()` would be sufficient.

I don't claim to be a Vitest expert though. I switched from Jest -> Vitest -> Bun, and experienced speed wins each time. But you're using newer versions of Jest and Vitest, so things may have changed. And these tests use JSDom, which I've never needed to use, which is slower than running tests on node.

The benefit of this PR, is not only does Vitest provide ESM support, it brings some future proofing because Vitest should eventually be rewritten in native code, given VoidZero's vision for a rust based end-to-end toolchain.

**Comment by birkskyum** [comment]
@jasongitmail , thank you very much for the review. It is a good idea to try to narrow down single files that might be impacting the numbers, and if beforeEach/All can help.

You can run Vitest with Bun? [I haven't had much success with that](https://github.com/oven-sh/bun/issues/4145). Unless the --bun flag is set, I believe it might be running vitest with node underneath still.

**Comment by jasongitmail** [comment]
I didn't run Vitest with Bun, I ported my tests to `bun:test` by changing the imports `import { expect, test } from "bun:test";`. In my case, it mostly just worked because `bun:test` aims to be Jest compatible. https://bun.sh/docs/cli/test  

It's an option, but I'm not suggesting that in this case b/c `maplibre-gl-js` probably has many contributors and this would require them to have Bun installed to run tests, and I'd guess you'll likely end up switching back to Vitest down the road for ecosystem alignment. But it could be fun to play with to see how fast tests can be.

**Comment by birkskyum** [comment]
@jasongitmail aah, that makes sense - the stats I've seen for bun:test are quite impressive, but I didn't even consider it when making this PR. As you say, it's requiring everyone to move to another runtime which would take a very convincing argument, and also until the day bun get a text-based lock-file, we'd be unable to use Dependabot, which we rely heavily on.

**Comment by birkskyum** [comment]
This describes some of the differences between the two runners

https://cookbook.marmicode.io/angular/why-vitest/

**Comment by HarelM** [comment]
My main concern is about coverage, I've spent a great deal of time to make the coverage work correctly for both render and unit tests. 
I still think it doesn't work well with latest jest (there's an open issue on it), but at least it's close to working correctly...

**Comment by birkskyum** [comment]
@HarelM , is this how you'd expect it to look for the unit tests? it's the `npm run test-unit-ci`

Screenshot 2024-11-21 at 23 24 52

**Comment by birkskyum** [comment]
This is the `npm run test-integration-ci`

Screenshot 2024-11-21 at 23 26 09

**Comment by HarelM** [comment]
I'd like to solve the current problem we have we coverage, this might be an opportunity to do so while removing the need to use all kind of crazy package we have today like monocart etc.
Can you configure the CI to upload the coverage reports (json files) to codecov so that I can take a look at things I know are not reported correctly today?

**Comment by birkskyum** [comment]
@HarelM , i've tried passing it the vitest json output for the Unit test suite to codecov - can you see if it's reasonable/?

**Comment by HarelM** [comment]
It reports 70k lines, originally it reported arond 40k.
I'll need to see if it reports comments and non functional code...

**Comment by birkskyum** [comment]
@HarelM , I made a mistake just before - it double counted the integration tests. should be resolved now

**Comment by birkskyum** [comment]
note: This still uses monocart for the render tests. Vitest only swaps the Jest tests (unit/integration), so monocart is still there.

We can probably swap it by leveraging the [Vitest Browser Mode](https://vitest.dev/guide/browser/) (experimental) - But if possible, I'd like to avoid it for this PR.

**Comment by birkskyum** [comment]
@HarelM the discrepancy in numbers is that vitest includes the .test.ts files. I'll try to exclude them.

**Comment by birkskyum** [comment]
@HarelM  I'm seeing the same number of tracked lines now.

**Comment by HarelM** [comment]
Should I be reviewing this? Did you close it by accident?

**Comment by birkskyum** [comment]
It was closed by accident - I think it's ready for review

File: src/source/rtl_text_plugin_main_thread.test.ts [hunk 1]
```diff
@@ -9,7 +10,7 @@ const rtlMainThreadPlugin = rtlMainThreadPluginFactory();
 
 describe('RTLMainThreadPlugin', () => {
     let server: FakeServer;
-    let broadcastSpy: jest.SpyInstance;
+    let broadcastSpy: ReturnType<typeof vi.spyOn>;
```
Comment (HarelM) [inline src/source/rtl_text_plugin_main_thread.test.ts:13]: Is there a better type for this?

**Comment by birkskyum** [comment]
And I went with the beta release, because I saw they've made a change i the 2.2.0-beta.1 related to including the coverageMap in the json file output:

https://github.com/vitest-dev/vitest/releases

File: src/source/rtl_text_plugin_main_thread.test.ts (hunk 1 above)
Comment (birkskyum) [reply src/source/rtl_text_plugin_main_thread.test.ts]: Yes, I've updated it to MockInstance

File: src/util/actor.test.ts [hunk 2]
```diff
@@ -94,11 +95,11 @@ describe('Actor', () => {
         expect(gotAbortSignal).toBeTruthy();
     });
 
-    test('cancel a request that must be queued will not call the method at all', async () => {
+    test('cancel a request that must be queued will not call the method at all', {retry: 3}, async () => {
```
Comment (HarelM) [inline src/util/actor.test.ts:98]: Why was this added?

File: src/util/is_char_in_unicode_block.test.ts [hunk 3]
```diff
@@ -1,14 +1,17 @@
+import {describe, test, expect} from 'vitest';
 import {unicodeBlockLookup} from './is_char_in_unicode_block';
 
 describe('unicodeBlockLookup', () => {
     test('each code block lookup function follows the same pattern', () => {
         for (const codeBlock in unicodeBlockLookup) {
             const lookup = unicodeBlockLookup[codeBlock];
-            const match = lookup.toString().match(/^\(char\) => char >= 0x([0-9A-F]{4,6}) && char <= 0x([0-9A-F]{4,6})$/);
+            const lookupString = lookup.toString();
+            console.log(`Checking code block: ${codeBlock}, function: ${lookupString}`);
```
Comment (HarelM) [inline src/util/is_char_in_unicode_block.test.ts:9]: Remove console.log.

File: src/util/web_worker_transfer.ts [hunk 4]
```diff
@@ -56,7 +56,7 @@ export function register<T extends any>(
     },
     options: RegisterOptions<T> = {}
 ) {
-    if (registry[name]) throw new Error(`${name} is already registered.`);
+    if (registry[name]) console.warn(`${name} is already registered.`);
```
Comment (HarelM) [inline src/util/web_worker_transfer.ts:59]: Why was this logic changed as part of this PR?

Comment (birkskyum) [reply src/util/web_worker_transfer.ts]: Screenshot 2024-11-22 at 14 10 29

**Comment by HarelM** [comment]
Coverage seems like it was increased, which seems correct for a last case I know of here, which I made sure the PR covers, but the report was lacking.
I hope this will be the end of the coverage problems.
If not by any other reason, this is a good enough reason for me to switch.
https://app.codecov.io/gh/maplibre/maplibre-gl-js/pull/4728/blob/src/util/dom.ts

As a side note, for consistency I would advise to move the vitest import to be the first import in all the tests files, but that my OCD I guess...

**Comment by HarelM** [comment]
Is the content of src/util/test/util.ts not needed anymore? I see that it is barely covered by tests, which it is intended for...

File: src/util/web_worker_transfer.ts (hunk 4 above)
Comment (birkskyum) [reply src/util/web_worker_transfer.ts]: I've put in place a check to avoid re-registration:
https://github.com/maplibre/maplibre-gl-js/pull/4728/commits/18bf9fff21b307dab82ecf9728be0c1a5032ff37

**Comment by birkskyum** [comment]
> Is the content of src/util/test/util.ts not needed anymore? I see that it is barely covered by tests, which it is intended for...

I see >100 uses of the StubMap, so we'd need a replacement of that

File: src/util/web_worker_transfer.ts [hunk 5]
```diff
@@ -56,7 +56,10 @@ export function register<T extends any>(
     },
     options: RegisterOptions<T> = {}
 ) {
-    if (registry[name]) throw new Error(`${name} is already registered.`);
+    if (registry[name]) {
+        console.warn(`${name} is already registered.`);
+        return;
+    }
```
Comment (birkskyum) [inline src/util/web_worker_transfer.ts:62]: Either the register() should only be called by explicit execution (not by just loading the file), or we can do this to let it handle multiple calls more gently

Comment (HarelM) [reply src/util/web_worker_transfer.ts:62]: How was it able to work before this change?

File: src/render/draw_symbol.test.ts [hunk 6]
```diff
@@ -17,13 +18,13 @@ import {Style} from '../style/style';
...
-jest.mock('../source/source_cache');
-jest.mock('../source/tile');
-jest.mock('../data/bucket/symbol_bucket');
-jest.mock('../symbol/projection');
-(symbolProjection.getPitchedLabelPlaneMatrix as jest.Mock).mockReturnValue(mat4.create());
+vi.mock('./painter');
+vi.mock('./program');
+vi.mock('../source/source_cache');
+vi.mock('../source/tile');
+vi.mock('../data/bucket/symbol_bucket');
```
Comment (birkskyum) [inline src/render/draw_symbol.test.ts:25]: @HarelM  This reads the file again, so it ends up being executed twice.

File: src/util/web_worker_transfer.ts (hunk 5 above)
Comment (birkskyum) [reply src/util/web_worker_transfer.ts:62]: I can imagine it's because Vitest is esm based, and imports modules a bit different.

**Review by HarelM — APPROVED** [review]
Looks good to me, I'm happy to see that the problem with the coverage is not fixed with this change.

**Comment by birkskyum** [comment]
>  I'm happy to see that the problem with the coverage is not fixed with this change.

Is there still a problem? This v5 that it's not reporting correctly in PRs, is that still ongoing after this?

**Comment by HarelM** [comment]
@birkskyum I've updated the repo and ran the tests locally. It takes around 14 seconds to complete the run. It used to take around 5 seconds with jest. Any ideas how to get it back to 5 seconds?

**Comment by birkskyum** [comment]
I do have some idea - I outlined them [here](https://osmus.slack.com/archives/C03TFH5NE83/p1732379052720919?thread_ts=1732347264.351569&cid=C03TFH5NE83).

**Comment by HarelM** [comment]
Fair enough, I'll continue monitoring the behavior.
